### ISSUE TITLE
[clang] Missing diagnostic on declarations of function & variable templates with the same name as a template parameter.

### ISSUE THREAD

**Issue by cor3ntin** [llvm/llvm-project#53925]
Clang does not diagnose C++ function templates and variable template declarations with the name of a template-parameter,
which from my reading is non-conforming.

```cpp
template <auto a> 
void a(); // no diagnostic, expected a redeclaration error

template <typename b> 
void b(); // no diagnostic, expected a redeclaration error

template <typename c> 
auto c = 0; // no diagnostic, expected a redeclaration error

template <auto d> 
auto d = 0; // no diagnostic, expected a redeclaration error

template <typename d> 
class d; // ok ( ill-formed, as expected)
```

Both GCC and MSVC complain about redeclarations (CE link https://godbolt.org/z/8o9ecbh95)

The C++ standard specifies
> The name of a template-parameter shall not be bound to any following declaration contained by the scope to which the template-parameter belongs.

http://eel.is/c++draft/temp.res#temp.local-6
